### PR TITLE
[Android] Replace `SharedPreferences` with `DataStore` Preferences

### LINKED ISSUE #628: Replace `SharedPreferences` with `DataStore` Preferences on Android
Hi!

I'm working on an implementation to replace the usage of `SharedPreferences` with `DataStore`. It's recommended by Google as seen in [Prefer Storing Data with Jetpack DataStore](https://android-developers.googleblog.com/2020/09/prefer-storing-data-with-jetpack.html), moreover security-wise there may be concerns about SharedPreferences being deprecated in the longer run.

Since `DataStore` is tailored to Kotlin, I plan to write the code for it in Kotlin, thus adding Kotlin support to the Android library project.

Hope everything about this is ok with the maintainers 🙏🏻

### PR DESCRIPTION
**Pull request by ovitrif**
Closes #628

This PR replaces the usage of `SharedPreferences` with [`DataStore`](https://developer.android.com/topic/libraries/architecture/datastore#kotlin).

## Why
Google is recommending in their official documentation to migrate to `DataStore` from `SharedPreferences`, and they also provided a very easy-to-use auto-migration tool that will move the data to the new `DataStore`, then clean up the entries from `SharedPreferences`.

Their official guide for [saving simple data with SharedPreferences](https://developer.android.com/training/data-storage/shared-preferences) displays an alert about it.

For security concerns as well, it is recommended to migrate to DataStore.

## How
Calls to read and write to DataStore are forced to run synchronously based on the example provided by the Android team in [Use DataStore in synchronous code](https://developer.android.com/topic/libraries/architecture/datastore#synchronous).

### Data Migration
[`SharedPreferencesMigration`](https://developer.android.com/reference/kotlin/androidx/datastore/migrations/SharedPreferencesMigration) is used to migrate the entries stored in the `RN_KEYCHAIN` shared prefs file to data store. Upon completion, the `SharedPreferences` get removed.

<details><summary>◐ Preview File System migration</summary>

| before | after |
| - | - |
| before | after |
</details>

### Testing
- ~~junit test were removed from the project~~
- Done manual testing via an app using React Native to verify migrations work seamlessly.
- Also fixed an issue with concurrent cipher operations, which worked great after the fix.

### CODE REVIEW

**Comment by ovitrif** [comment]
@oblador Any idea why the error about missing manifest in the Unit Tests?
Might have been a side-effect of adding Kotlin 🤷🏻

File: build.gradle.kts [hunk 1]
```diff
@@ -10,6 +10,8 @@ buildscript {
     set("targetSdkVersion", 31)
     set("buildToolsVersion", "31.0.0")
   }
+  val kotlinVersion by extra("1.8.0")
```
Comment (enahum) [inline build.gradle.kts:13]: If the root project uses kotlin this should use that version instead, ex my project is using 1.9.22

**Review by enahum — COMMENTED** [review]
This would be a good addition to the library

File: build.gradle.kts (hunk 1 above)
Comment (ovitrif) [reply build.gradle.kts:13]: Thanks, yeah, good version

**Comment by DorianMazur** [comment]
@ovitrif Can you update this PR?

File: build.gradle.kts (hunk 1 above)
Comment (ovitrif) [reply build.gradle.kts]: kotlin version is now managed from `gradle.properties`, as I've replayed the changes on top of the latest version 👍🏻

**Comment by ovitrif** [comment]
> @ovitrif Can you update this PR?

Updated sir, also now I could make use of the `coroutineScope` you introduced with the Kotlin migration, and hot reload works again 🎉

**Comment by ovitrif** [comment]
@DorianMazur Please let me know if you're suggesting other updates 👍🏻

**Comment by ovitrif** [comment]
FYI I found an issue that is not related to the datastore update but highlighted by it:
```java
Unknown error with alias: pin, error: Cannot update the same operation concurrently. (Ask Gemini)
com.oblador.keychain.exceptions.CryptoFailedException: Unknown error with alias: pin, error: Cannot update the same operation concurrently.
	at com.oblador.keychain.cipherStorage.CipherStorageKeystoreAesCbc.encrypt(CipherStorageKeystoreAesCbc.kt:101)
	at com.oblador.keychain.KeychainModule$setGenericPassword$1.invokeSuspend(KeychainModule.kt:210)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108)
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697)
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684)
Caused by: java.lang.IllegalThreadStateException: Cannot update the same operation concurrently.
	at android.security.KeyStoreOperation.handleExceptions(KeyStoreOperation.java:74)
	at android.security.KeyStoreOperation.update(KeyStoreOperation.java:118)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer$MainDataStream.update(KeyStoreCryptoOperationChunkedStreamer.java:222)
	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.update(KeyStoreCryptoOperationChunkedStreamer.java:156)
	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineUpdate(AndroidKeyStoreCipherSpiBase.java:436)
	at javax.crypto.Cipher.update(Cipher.java:1741)
	at javax.crypto.CipherOutputStream.write(CipherOutputStream.java:158)
	at javax.crypto.CipherOutputStream.write(CipherOutputStream.java:144)
	at com.oblador.keychain.cipherStorage.CipherStorageBase.encryptString(CipherStorageBase.kt:369)
	at com.oblador.keychain.cipherStorage.CipherStorageKeystoreAesCbc.encryptString(CipherStorageKeystoreAesCbc.kt:238)
	at com.oblador.keychain.cipherStorage.CipherStorageKeystoreAesCbc.encrypt(CipherStorageKeystoreAesCbc.kt:97)
	at com.oblador.keychain.KeychainModule$setGenericPassword$1.invokeSuspend(KeychainModule.kt:210) 
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33) 
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:108) 
	at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:584) 
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:793) 
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:697) 
	at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:684) 
	Suppressed: java.lang.IllegalThreadStateException: Cannot update the same operation concurrently.
		at android.security.KeyStoreOperation.handleExceptions(KeyStoreOperation.java:74)
		at android.security.KeyStoreOperation.finish(KeyStoreOperation.java:132)
		at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer$MainDataStream.finish(KeyStoreCryptoOperationChunkedStreamer.java:228)
		at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.doFinal(KeyStoreCryptoOperationChunkedStreamer.java:181)
		at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:626)
		at javax.crypto.Cipher.doFinal(Cipher.java:1957)
		at javax.crypto.CipherOutputStream.close(CipherOutputStream.java:210)
```

Working on a fix.

**Comment by DorianMazur** [comment]
Great work @ovitrif Thank you! I'll check out that PR and merge it, probably this week.

**Pull request merged**